BUGFIX: Allow non-entity asset variants

bwaidelich opened the pull request:
Allow implementations of the `AssetVariantInterface` to not be doctrine entities.

Without this fix, all `AssetRepository::find*()` calls lead to an exception when a class implements the interface but is not an entity (for example because it is just used to generate dummy images for Monocle).

## Steps to reproduce

1. Implement the interface with a new custom class like

```php
final class DummyImageVariant implements AssetVariantInterface {
    // ...
}
```

2. Trigger `AssetRepository::find*()` method, eg. by invoking `./ flow media:removeunused`

### Expected result

No error, the custom implementation should be ignored

### Actual result

Exception:

```
Failure while fetching class schema class "Some\Package\DummyImageVariant": No class schema found for "Some\Package\DummyImageVariant".

  Type: Doctrine\ORM\Mapping\MappingException
  Code: 1542792708
  File: Data/Temporary/Development/Cache/Code/Flow_Object_Classes/Neos_Flow_Persist
        ence_Doctrine_Mapping_Driver_FlowAnnotationDriver.php
  Line: 187
```

bwaidelich committed:
BUGFIX: Allow non-entity asset variants

Allow implementations of the `AssetVariantInterface` to not be doctrine entities.

Without this fix, all `AssetRepository::find*()` calls lead to an exception when a class implements the interface
but is not an entity (for example because it is just used to generate dummy images for Monocle).

1. Implement the interface with a new custom class like

```php
final class DummyImageVariant implements AssetVariantInterface {
    // ...
}
```

2. Trigger `AssetRepository::find*()` method, eg. by invoking `./ flow media:removeunused`

No error, the custom implementation should be ignored

Exception:

```
Failure while fetching class schema class "Some\Package\DummyImageVariant": No class schema found for "Some\Package\DummyImageVariant".

  Type: Doctrine\ORM\Mapping\MappingException
  Code: 1542792708
  File: Data/Temporary/Development/Cache/Code/Flow_Object_Classes/Neos_Flow_Persist
        ence_Doctrine_Mapping_Driver_FlowAnnotationDriver.php
  Line: 187
```

mhsdesign commented on Neos.Media/Classes/Domain/Repository/AssetRepository.php:
```diff
@@ -284,7 +284,9 @@ protected function addAssetVariantFilterClause(Query $query): void
         $variantClassNames = $this->reflectionService->getAllImplementationClassNamesForInterface(AssetVariantInterface::class);
         foreach ($variantClassNames as $variantClassName) {
-            $queryBuilder->andWhere('e NOT INSTANCE OF ' . $variantClassName);
+            if (is_a($variantClassName, AssetInterface::class)) {
```
i think we should use 

`in_array($interface, class_implements($value), true)`

@nezaniel pointed out to me the flaws of `is_a` but i forgot ^^

bwaidelich commented:
~Please don't merge, the fix no longer works with Neos 8.3~

fixed again :)

bwaidelich committed:
Update AssetRepository.php

The fix no longer worked with 8.3 due to some other changes, so it was adjusted
